internal/controllers/topology/cluster/desired_state.go

```diff
@@ -772,86 +771,72 @@ func computeMachineDeploymentVersion(s *scope.Scope, machineDeploymentTopology c
-		return currentVersion, nil
+		s.UpgradeTracker.MachineDeployments.MarkPendingUpgrade(currentMDState.Object.Name)
+		return currentVersion
 	}
 
-	// Return early if we are not allowed to upgrade the machine deployment.
-	if !s.UpgradeTracker.MachineDeployments.AllowUpgrade() {
+	// Return early if the AfterControlPlaneUpgrade hook returns a blocking response.
+	if s.HookResponseTracker.IsBlocking(runtimehooksv1.AfterControlPlaneUpgrade) {
 		s.UpgradeTracker.MachineDeployments.MarkPendingUpgrade(currentMDState.Object.Name)
-		return currentVersion, nil
+		return currentVersion
 	}
 
-	// If the control plane is being created (current control plane is nil), do not perform
-	// any machine deployment upgrade in this case.
-	// Return the current version of the machine deployment.
-	// NOTE: this case should never happen (upgrading a MachineDeployment) before creating a CP,
-	// but we are implementing this check for extra safety.
-	if s.Current.ControlPlane == nil || s.Current.ControlPlane.Object == nil {
+	// Return early if the upgrade concurrency is reached.
+	if s.UpgradeTracker.MachineDeployments.UpgradeConcurrencyReached() {
 		s.UpgradeTracker.MachineDeployments.MarkPendingUpgrade(currentMDState.Object.Name)
-		return currentVersion, nil
+		return currentVersion
 	}
 
-	// If the current control plane is upgrading, then do not pick up the desiredVersion yet.
+	// Return early if the Control Plane is not stable. Do not pick up the desiredVersion yet.
 	// Return the current version of the machine deployment. We will pick up the new version after the control
 	// plane is stable.
-	cpUpgrading, err := contract.ControlPlane().IsUpgrading(s.Current.ControlPlane.Object)
-	if err != nil {
-		return "", errors.Wrap(err, "failed to check if control plane is upgrading")
-	}
-	if cpUpgrading {
+	if !isControlPlaneStable(s) {
 		s.UpgradeTracker.MachineDeployments.MarkPendingUpgrade(currentMDState.Object.Name)
-		return currentVersion, nil
+		return currentVersion
+	}
+
+	// Control plane and machine deployments are stable.
+	// Ready to pick up the topology version.
+	s.UpgradeTracker.MachineDeployments.MarkUpgrading(currentMDState.Object.Name)
+	return desiredVersion
+}
+
+// isControlPlaneStable returns true is the ControlPlane is stable.
+func isControlPlaneStable(s *scope.Scope) bool {
+	// If the current control plane is upgrading it is not considered stable.
+	if s.UpgradeTracker.ControlPlane.IsUpgrading {
+		return false
 	}
 
 	// If control plane supports replicas, check if the control plane is in the middle of a scale operation.
-	// If the current control plane is scaling, then do not pick up the desiredVersion yet.
-	// Return the current version of the machine deployment. We will pick up the new version after the control
-	// plane is stable.
+	// If the current control plane is scaling then it is not considered stable.
 	if s.Blueprint.Topology.ControlPlane.Replicas != nil {
```

Comment:
```suggestion
```
nit: I think we can drop this here. This check is part of the earlier calculation if the CP is scaling or not. We don't have to consider it (aka "calculate") here again

(+ update/simplify godoc)